src/app/modules/item/services/threads.service.ts

```diff
@@ -44,37 +98,115 @@ interface TokenData {
   canWatchParticipant: boolean,
 }
 
-type Action = 'open-thread' | 'close-thread' | 'follow';
+type ThreadAction =
+  | { action: 'open-thread', history: ActivityLog[] }
+  | { action: 'close-thread' }
+  | { action: 'follow' }
+  | { action: 'unfollow' }
+  | { action: 'thread-status' }
+  | { action: 'send-message', message: string };
 
 @Injectable({
   providedIn: 'root',
 })
-export class ThreadService {
+export class ThreadService implements OnDestroy {
+
+  events$: Observable<ThreadEventMessage[]>;
+  status$: Observable<ThreadStatusMessage['status']>;
 
+  private newEvents$: Observable<ThreadEventMessage[]>;
+  private clearEvents$ = new Subject<void>();
   private tokenData?: TokenData;
   private socket: WebSocketSubject<unknown>;
-  messages$: Observable<ThreadEvent[]>;
+  private incomingMessages$: Observable<ThreadMessage[]>;
+  private subscriptions: Subscription[] = [];
 
-  constructor() {
+  constructor(
+    private activityLogService: ActivityLogService,
+  ) {
     if (!appConfig.forumServerUrl) throw new Error('cannot instantiate threads service when no forum server url');
     this.socket = webSocket(appConfig.forumServerUrl);
-    this.messages$ = this.socket.pipe(
-      decodeSnakeCase(D.array(threadEventDecoder)),
-      scan((oldEvents, newEvents) => [ ...oldEvents, ...newEvents ]),
+
+    this.incomingMessages$ = this.socket.pipe(
+      decodeSnakeCase(D.array(threadMessageDecoder)),
       catchError(() => EMPTY), // ignore undecoded messages
     );
+
+    this.newEvents$ = this.incomingMessages$.pipe(map(messages => messages.filter(isThreadEventMessage)));
+
+    this.events$ = this.clearEvents$.pipe(
+      startWith(undefined),
+      switchMap(() => this.newEvents$.pipe(scan((oldEvents, newEvents) => [ ...oldEvents, ...newEvents ]))),
+      map(messages => messages.sort((a, b) => a.time.valueOf() - b.time.valueOf())), // sort by date ascending
+    );
+
+    this.status$ = this.incomingMessages$.pipe(
+      map(messages => messages.find(isThreadStatusMessage)?.status),
+      filter(isNotUndefined),
+      distinctUntilChanged(),
```

Comment:
You would then have a result of type array, which is not what I want :thinking:

Reply:
Ok, sorry.

And there is a guarantee there is only one single status message among the messages?

Reply:
This part of the code is no longer valid since thread-status is now an http service.
To answer your question though: in the array we keep (which contains the newest events), there might be multiple status-related events (open/close) or none (when the last open/close event is older than the last 20 events at init).
1. If there are thread-status-related events in our array, we consider the last one.
2. If there are none, we ask the thread-status http service.
3. If we start with no thread-status-related event in our array and receive one afterwards, case 1 applies.

Anyhow, the thread status is unique at a given time.